### PR TITLE
Replace direct `cudaMemcpyAsync` calls with utility functions (within `/include`)

### PR DESCRIPTION
**Pull request by vuule**
## Description
Replaced the calls to `cudaMemcpyAsync` with the new `cuda_memcpy`/`cuda_memcpy_async` utility, which optionally avoids using the copy engine.

Also took the opportunity to use `cudf::detail::host_vector` and its factories to enable wider pinned memory use.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cudf/detail/get_value.cuh [hunk 1]
```diff
@@ -48,11 +49,8 @@ T get_value(column_view const& col_view, size_type element_index, rmm::cuda_stre
   CUDF_EXPECTS(data_type(type_to_id<T>()) == col_view.type(), "get_value data type mismatch");
   CUDF_EXPECTS(element_index >= 0 && element_index < col_view.size(),
                "invalid element_index value");
-  T result;
-  CUDF_CUDA_TRY(cudaMemcpyAsync(
-    &result, col_view.data<T>() + element_index, sizeof(T), cudaMemcpyDefault, stream.value()));
-  stream.synchronize();
```
Comment (davidwendt) [inline cpp/include/cudf/detail/get_value.cuh:54]: This code is removed in https://github.com/rapidsai/cudf/pull/17520

Comment (vuule) [reply cpp/include/cudf/detail/get_value.cuh:54]: This comment somehow moved to a different file :D
Should be resolved now, updated the branch and resolved conflicts.

File: cpp/include/cudf/detail/get_value.cuh [hunk 2]
```diff
@@ -48,11 +49,8 @@ T get_value(column_view const& col_view, size_type element_index, rmm::cuda_stre
...
-  return result;
+  return cudf::detail::make_host_vector_sync(
+    device_span<T const>{col_view.data<T>() + element_index, 1}, stream)[0];
```
Comment (davidwendt) [inline cpp/include/cudf/detail/get_value.cuh:53]:
```suggestion
    device_span<T const>{col_view.data<T>() + element_index, 1}, stream).front();
```

**Review by davidwendt — APPROVED** [review]

File: cpp/include/cudf/detail/get_value.cuh [hunk 3]
```diff
@@ -48,11 +49,8 @@ T get_value(column_view const& col_view, size_type element_index, rmm::cuda_stre
   CUDF_EXPECTS(data_type(type_to_id<T>()) == col_view.type(), "get_value data type mismatch");
   CUDF_EXPECTS(element_index >= 0 && element_index < col_view.size(),
                "invalid element_index value");
-  T result;
-  CUDF_CUDA_TRY(cudaMemcpyAsync(
-    &result, col_view.data<T>() + element_index, sizeof(T), cudaMemcpyDefault, stream.value()));
-  stream.synchronize();
-  return result;
+  return cudf::detail::make_host_vector_sync(
+    device_span<T const>{col_view.data<T>() + element_index, 1}, stream).front();
```
Comment (nvdbaranec) [inline cpp/include/cudf/detail/get_value.cuh]: Quibble: In some alternate universe where pinned memory isn't being used and we drop back to the pageable resource, this is probably somewhat worse than what's there now.  I imagine that's not a case we really care about though.

**Review by nvdbaranec — APPROVED** [review]

File: cpp/include/cudf/detail/get_value.cuh (hunk 3 above)
Comment (vuule) [reply cpp/include/cudf/detail/get_value.cuh]: That's true. FWIW, we did a full benchmark run with a change like this in the `device_scalar`, and there was no negative impact from involving the heap unnecessarily.

**Review by karthikeyann — APPROVED** [review]

**Pull request merged**